### PR TITLE
SALTO-791: add custom group key function to adapter api

### PR DESCRIPTION
**Pull request by ori-moisis**

### CODE REVIEW

File: packages/core/src/core/plan/group.ts [hunk 1]
```diff
@@ -42,10 +43,39 @@ export const getOrCreateGroupLevelChange = (group: Group<Change>, beforeElements
...
+
+  const changesPerAdapter = collections.iterable.groupBy(
+    wu(changes.keys()).map(id => ({ id, change: changes.getData(id) })),
+    ({ change }) => getChangeElement(change).elemID.adapter,
+  )
+
+  const changeGroupIds = wu(changesPerAdapter.entries())
+    .filter(([adapterName]) => adapterName in customGroupIdFunctions)
+    .map(([name, adapterChanges]) => (
+      customGroupIdFunctions[name](new Map(adapterChanges.map(({ id, change }) => [id, change])))
```
Comment (omrilit) [inline packages/core/src/core/plan/group.ts:37]: nit: you could create adapterChanges as a `Map` at first place and then avoid this transformation.

File: packages/core/src/core/plan/filter.ts [hunk 2]
```diff
@@ -103,8 +102,22 @@ export const filterInvalidChanges = async (
...
+      if (isField(elem)) {
+        const validParent = validAfterElementsMap[elem.parent.elemID.getFullName()] as ObjectType
+        return validParent?.fields?.[elem.name]
+      }
+      return validAfterElementsMap[elem.elemID.getFullName()]
+    }
+    const replaceAfterElement = <T extends DiffNode<ChangeDataType>>(change: T): T => {
+      if (isAdditionOrModificationDiff(change)) {
+        const validAfter = getValidAfter(getChangeElement(change)) ?? change.data.after
+        return { ...change, data: { ...change.data, after: validAfter ?? change.data.after } }
```
Comment (omrilit) [inline packages/core/src/core/plan/filter.ts:116]: You already have the `?? change.data.after` fallback one row above. I think there is no need for it here

Comment (ori-moisis) [reply packages/core/src/core/plan/filter.ts:116]: you are very correct

File: packages/core/src/core/plan/group.ts (hunk 1 above)
Comment (ori-moisis) [reply packages/core/src/core/plan/group.ts:37]: I could not figure out a way to do this in one go (without writing a very complicated alternative to `groupBy` in lowerdash)

**Pull request merged**